### ISSUE TITLE
bfloat16 running 4x slower than fp32

### ISSUE THREAD

**Issue by ahaliassos** [PyTorchLightning/pytorch-lightning#11933]
## 🐛 Bug

I'm training a hybrid Resnet18+Conformer model using A100 GPUs. I've used both fp16 and fp32 precision to train the model and things work as expected: fp16 uses less memory and runs faster than when using fp32. However, when I try to switch to bf16 through `Trainer(precision='bf16')`, training speed is dramatically lower than with fp32, from ~1.3s/it to ~5.5s/it. The only difference between the runs is the `precision` argument, where I used `'16'` for fp16, `'32'` for fp32, and `'bf16'` for bfloat16. The slowness of bf16 is experienced when using only 1 GPU, 8 GPUs (single node), or 32 GPUs (multiple nodes). 

The [docs](https://pytorch-lightning.readthedocs.io/en/latest/advanced/mixed_precision.html) state 

> BFloat16 is also experimental and may not provide significant speedups or memory improvements, offering better numerical stability.

But shouldn't it work faster than fp32 on A100 GPUs? Any help would be hugely appreciated.

### Expected behavior

For bf16 to work faster than fp32 and probably slower than fp16.

### Environment

* CUDA:
	- GPU:
		- A100-SXM4-40GB
	- available:         True
	- version:           11.3
* Packages:
	- numpy:             1.21.2
	- pyTorch_debug:     False
	- pyTorch_version:   1.10.2
	- pytorch-lightning: 1.5.9
	- tqdm:              4.62.3
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.8.12

### Additional context

Although speed is dramatically lower with bf16 than with fp32, the GPU memory allocation is lower (~55% vs ~95%), as expected.

**Comment by ananthsub** [comment]
Do you have a pointer to your code for us to investigate?

**Comment by joshua840** [comment]
Hi @ananthsub, I have the same problem and my trial code is available here: (https://github.com/PyTorchLightning/pytorch-lightning/issues/12600). In my case, fp32 is about x1.3 faster than bf16 and fp16 is ~x1.9 faster than fp32. 
I tested with A100 and A5000, with PyTorch 1.11 and Lightning 1.6.

**Comment by carmocca** [comment]
@ahaliassos It could be caused by the specific operations you have in your `LightningModule`. AFAIK torch's `einsum` is not the most performant.

Do you see experience speed-downs with simpler modules?

**Comment by joshua840** [comment]
Hi @carmocca, I removed the `einsum` operation parts from (https://github.com/PyTorchLightning/pytorch-lightning/issues/12600) and checked the training speeds for each precision.

The training speeds are as follows
- fp16: 13.10it/s
- fp32: 8.78it/s
- bf16: 4.55it/s

Briefly, my code calculates the cross-entropy loss during training and updates the ResNet18 model with ImageNet dataset.

**Comment by carmocca** [comment]
We will try to repro using this script https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/domain_templates/imagenet.py and report back

**Comment by krshrimali** [comment]
Hi, @joshua840 - Thank you so much for reporting this issue! Acknowledging that we were able to reproduce this issue on our servers, and this is something I'll debug further. We will get back to you as soon as possible, whenever we have some pointers on this. Thank you again!

**Comment by krshrimali** [comment]
Hi, so I was finally able to find some time to benchmark for possible precisions (apologies for the delay).

| Precision        | PyTorch (iters/s) | PyTorch Lightning (iters/s) |
| ------------- |:-------------:      | :-----:                                   |
| `float16`        | 3.92                     | 4.27                                       |
| `float32`        | 3.77                     | 4.22                                       |
| `bfloat16`      | 4.08                     | 2.5                                         |

**Note**:

1. Tested on: Image Net dataset
2. Scripts used:
    * PL: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/domain_templates/imagenet.py
    * PyTorch: https://github.com/pytorch/examples/blob/main/imagenet/main.py (used AMP for float16 and bf16 precisions)
3. Tested on ~2500 iterations. Numbers are averaged out.

**Analysis**:

- From the numbers, it looks like PyTorch Lightning is around 1.6 times slower than PyTorch for bfloat16 precision, while for other precisions - there doesn't seem to be a huge difference. PL performs a little faster, but I assume it will be near to equal if tested for >1 epochs.
- Also, I agree that bf16 is running slower than fp32 in PyTorch Lightning, which is not expected.

We'll try investigating why performance is slower for bf16 in PyTorch Lightning and will get back to this issue once we have some numbers. :) Thank you for your patience and for reporting this issue to us!

cc: @carmocca @SeanNaren @Borda for visibility! :)

**Comment by SeanNaren** [comment]
Working with @krshrimali to try to investigate! 

I'm excluding Lightning and using the [pytorch imagenet example](https://github.com/pytorch/examples/blob/main/imagenet/main.py) to gather the below results on an A100. I ran for more than 2500 iterations.

| Precision      | seconds per iteration |
|----------------|-----------------------|
| FP32           | 0.103                 |
| FP32 (No TF32) | 0.160                 |
| FP16           | 0.079                 |
| BFloat16       | 0.415                 |

I also tried the latest NGC container which improved results marginally but showed a massive slowdown for BFloat16. I also tried `resnet101` and still showed a slowdown.

I also tried adding `torch.cuda.synchronize()` before measurements were taken, and a negligible difference was made.

I think it will be worth checking other model types with a minimal script to see if we see the same slow down without PyTorch Lightning before investigating.

Note: no TF32 meant adding this:

```python
# The flag below controls whether to allow TF32 on matmul. This flag defaults to True.
torch.backends.cuda.matmul.allow_tf32 = False

# The flag below controls whether to allow TF32 on cuDNN. This flag defaults to True.
torch.backends.cudnn.allow_tf32 = False
```

**Comment by blefaudeux** [comment]
ping @SeanNaren @ananthsub, any news on that ?

**Comment by blefaudeux** [comment]
> ping @SeanNaren @ananthsub, any news on that ?

ok, synced offline and seems that it's not a priority :(

**Comment by SeanNaren** [comment]
Here are my results running a pytorch-lightning minGPT script ([here](https://gist.github.com/SeanNaren/c9b69dc50d13e9e359ba4c49a1727c59)). This was a transformer model, with the below cmd:

```
python min_gpt.py --accelerator gpu --devices 1 --precision X
```

| Precision               | Iterations/s | nvidia-smi Memory (MiB) |
|-------------------------|--------------|-------------------------|
| 16                      | 2.26         | 30038                   |
| bf16                    | 2.24         | 30038                   |
| bf16 + model.bfloat16() | 2.40         | 24550                   |

I also used `torch.set_float32_matmul_precision('medium')` which gave roughly the same results.

We can see that the performance is roughly the same when using AMP, but when converting the model we get a small speedup + a significant memory drop as expected.

**It seems for transformer architectures BFloat16 works as expected**.

I found this [forum post](https://discuss.pytorch.org/t/bfloat16-has-worse-performance-than-float16-for-conv2d/154373) and [this comment](https://github.com/pytorch/pytorch/issues/57707#issuecomment-1166656767) describing slowdowns when using Conv2d + Bfloat16.

The model type in OP's post is based on convs, and the above reference issue could explain the reason why we see a 4x slowdown.

**Comment by blefaudeux** [comment]
That's great, thanks so much @SeanNaren ! The memory benefits are huge, for a similar speed it's still very useful. Cc @matthieutphr @quentindrx

**Comment by foreveronehundred** [comment]
I still have some problems regarding to the training speed of bfloat16 when enabling autocast.
Could anyone see the problem here?
https://discuss.pytorch.org/t/performance-training-speed-of-autocast-bfloat16/158183
